Edits for JavaScript 2021 chapter (#2614)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/javascript.md

---

## Introduction

The speed and consistency in which the JavaScript language has evolved with over the past years is tremendous. While in the past it was used primarily for client-side, it has taken a very important and respected place in the world of building services and server-side tools. JavaScript has evolved to a point where it is not only possible to create faster applications but also to <a hreflang="en" href="https://blog.stackblitz.com/posts/introducing-webcontainers/">run servers within browsers</a>.

There is a lot that happens in the browser when rendering the application, from downloading JavaScript to parsing, compiling and executing it. Let's start with that first step and try to understand how much JavaScript is actually requested by pages.

## How much JavaScript do we load?

To measure is the key towards improvement, they say. To improve the usage of JavaScript in our applications, we need to measure how much of the JavaScript being shipped is actually required. Let's dig in to understand the distribution of JavaScript bytes per page, considering what a major role it plays in the web setup.

> {# TODO - "To measure is the key towards improvement, they say." - is this a quotation or a reference to something? the phrasing implies it #}

Author: @NishuGoel

Reviewer: It's a reference to [this quote](https://www.goodreads.com/quotes/632992-measurement-is-the-first-step-that-leads-to-control-and) and variations of it. Didn't credit the quote author per se as its not really the usage of that quote but understanding in general.

Reviewer: @shantsis Hey, instead of 
"they say that, to measure...", could you put it as:

They say, "To measure is the key towards..."

Just sounds a bit better.

Author: Done!

### JavaScript requests per page

One of the contributing factors towards slow rendering of the web page could be the requests made on the page, especially when they are blocking. It's therefore of interest to look at the number of JavaScript requests made per page on both desktop and mobile devices.

> {# TODO - blocking something or just blocking? that's a lack of domain knowledge question :) #}

Author: @NishuGoel

Reviewer: especially when they are blocking - meaning especially when they are blocking [requests]

      </tr>
      <tr>
        <td>`nomodule`</td>
        <td class="numeric">3.9%</td>
        <td class="numeric">3.9%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(
    caption="Distribution of differential loading usage on desktop and mobile clients.",
    sheets_gid="1261294750",
    sql_file="module_and_nomodule.sql"
  ) }}</figcaption>
</figure>

4.6% of desktop pages use the `type="module"` attribute, whereas only 3.9% of mobile pages use `type="nomodule"`. This could be due to the fact that the mobile dataset being [much larger](./methodology#websites) contains more "long-tail" websites that might not be using the latest features.

It is important to note that with the <a hreflang="en" href="https://docs.microsoft.com/en-us/lifecycle/announcements/internet-explorer-11-support-end-dates">end of support for IE 11 browser</a>, differential loading is less applicable because evergreen browsers support modern JavaScript syntax. The Angular framework, for example, <a hreflang="en" href="https://github.com/angular/angular/issues/41840">removed support for legacy browsers in Angular v13</a>, which was released November 2021.

### `async` and `defer`

> JavaScript loading could be render-blocking unless it is specified as asynchronous or deferred. This is one of the contributing factors to slow performance, as often times JavaScript (or at least some of it) is needed for the initial render.

Reviewer 2: ```suggestion
JavaScript loading could be render-blocking unless it is specified as asynchronous or deferred. This is one of the contributing factors to slow performance, as oftentimes JavaScript (or at least some of it) is needed for the initial render.
```

### `async` and `defer`

JavaScript loading could be render blocking unless it is specified as deferred or async. This is one of the contributing factors to the slow website loading as often times JavaScript (or at least some of it) is needed for the initial render.

However, loading the JavaScript asynchronously or deferring it helps in some ways to improve this experience. The `async` and `defer` attribute load the scripts asynchronously. The scripts with `async` attribute are executed irrespective of the order in which they are defined, however, `defer` executes the scripts only after the document is completely parsed and ensures their execution will take place in the specified order. We will see the statistics for how many pages actually specify these attributes for the JavaScript requested in the browser.

> {# TODO - not sure how to rephrase "We will see the statistics" to say the stats are right below this paragraph #}

Author: @tunetheweb or @NishuGoel any thoughts?

Reviewer: Could simply say, "Let's look at how many pages......."

<!-- End Piwik Code -->
```

So, what is Piwik? According to its Wikipedia entry:

<figure>

> <blockquote>Matomo, formerly Piwik, is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL webserver. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools&hellip;</blockquote>

Reviewer 2: ```suggestion
<blockquote>Matomo, formerly Piwik, is a free and open source web analytics application developed by a team of international developers, that runs on a PHP/MySQL web server. It tracks online visits to one or more websites and displays reports on these visits for analysis. As of June 2018, Matomo was used by over 1,455,000 websites, or 1.3% of all websites with known traffic analysis tools&hellip;</blockquote>
```

This is where one has to put their 'performance-nerd' cap on, and ensure that using third-party scripts <a hreflang="en" href="https://csswizardry.com/2017/07/performance-and-resilience-stress-testing-third-parties/">doesn't result in losing control over the performance of the page,  or the main thread getting bloated</a> with too much JavaScript being transferred by <a hreflang="en" href="https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/loading-third-party-javascript">loading these scripts better</a>.

> {# TODO - 'performance-nerd' the quotes are rendering weirdly in the staged link #}

Author: @tunetheweb any thoughts?

Reviewer 3: Good spot. Should be double quotes anyway and that fixes it. Can you update?

  )
}}

> The median desktop pages requests 11 third-party resources, compared to 10 first-party requests. Irrespective of the <a hreflang="en" href="https://css-tricks.com/potential-dangers-of-third-party-javascript/">performance and reliability risks</a> that third-party resources may bring, both desktop and mobile pages consistently seem to favor third-party scripts. This effect could be due to the <a hreflang="en" href="https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript">useful interactivity features</a> that third-party scripts give to the web.

Reviewer 2: ```suggestion
The median desktop page requests 11 third-party resources, compared to 10 first-party requests. Irrespective of the <a hreflang="en" href="https://css-tricks.com/potential-dangers-of-third-party-javascript/">performance and reliability risks</a> that third-party resources may bring, both desktop and mobile pages consistently seem to favor third-party scripts. This effect could be due to the <a hreflang="en" href="https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript">useful interactivity features</a> that third-party scripts give to the web.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1265752379&format=interactive",
  sheets_gid="1107832831",
  sql_file="resource-hints-prefetch-preload-percentage.sql"
  )
}}

The median desktop page loads one JavaScript resource with the `preload` hint and two JavaScript resources with the `prefetch` hint.

<figure>
  <table>
    <thead>
      <tr>
        <th>Hint</th>
        <th>2020</th>
        <th>2021</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>`preload`</td>
        <td class="numeric">1</td>
        <td class="numeric">1</td>
      </tr>
      <tr>
        <td>`prefetch`</td>
        <td class="numeric">3</td>
        <td class="numeric">2</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(
    caption="Year-over-year comparison of the median number of `preload` and `prefetch` hints for JavaScript resources per mobile page.",
    sheets_gid="1107832831",
    sql_file="resource-hints-prefetch-preload-percentage.sql"
  ) }}</figcaption>
</figure>

> While the median number of `preload` hints per mobile page has stayed the same, the number of `prefetch` hints has decreased from three to two per page. Note that at the median, these resuts are identical for both mobile and desktop pages.

Reviewer 2: ```suggestion
While the median number of `preload` hints per mobile page has stayed the same, the number of `prefetch` hints has decreased from three to two per page. Note that at the median, these results are identical for both mobile and desktop pages.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1269923134&format=interactive",
  sheets_gid="821396474",
  sql_file="compression_method_by_3p.sql"
  )
}}

59.1% of third-party scripts on mobile pages are gzipped and 29.6% are compressed with Brotli. Looking at first-party scripts, these are 51.7% with Gzip compression but only 32.0% with Brotli. There are still 11.3% of third-party scripts that do not have any compression method defined.

{{ figure_markup(
  image="uncompressed-first-third-party.png",
  caption="Uncompressed resources for first party vs third party.",
  description="Bar chart showing the percent of the uncompressed resources for first party vs third party. 90% of uncompressed third party JavaScript resources are < 5 KB, and uncompressed first-party resources less than 10kb are only 8%.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1043138204&format=interactive",
  sheets_gid="1942216454",
  sql_file="compression_none_by_bytes.sql"
  )
}}

> 90% of uncompressed third-party JavaScript resources are less than 5 KB, though first-party requests trail a bit. This may help explain why so many JavaScript resources go uncompressed. Due to the diminishing returns of compressing small resources, a small script may cost more in terms of the resource consumption of server-side compression and client-side decompression than the performance benefits of saving of a few bytes over the network.

Reviewer 2: ```suggestion
90% of uncompressed third-party JavaScript resources are less than 5 KB, though first-party requests trail a bit. This may help explain why so many JavaScript resources go uncompressed. Due to the diminishing returns of compressing small resources, a small script may cost more in terms of the resource consumption of server-side compression and client-side decompression than the performance benefits of saving a few bytes over the network.
```

### Compression

Most of the compressed resources use either gzip compression, brotli (br) compression, or a compression technique that is not set.

> {# TODO - should we have links for gzip and brotli? i think other chapters do #}

Author: @NishuGoel

Reviewer: Sure.
Brotli - https://github.com/google/brotli
Gzip - https://www.gnu.org/software/gzip/manual/gzip.html

Could you update please?

It is found that only 0.1% of mobile pages use the SourceMap response header on script resources. One reason for this extremely small percentage could be that not many sites choose to put their original source code in production through the source map.

An analysis of how many sites actually send the source map header on their first-party or third-party scripts shows that 0.1% of the JavaScript requests that include a SourceMap header on mobile are for first-party scripts.

> {# TODO That seems surprising to me. I would have though it was more common to be from third-parties since the likes of query and Bootstrap include source map files. Any thoughts on why this is the reverse? Or am I just wrong about this expectation? Nishu - These results are confusing to me. I reran the query and didn't find 98% anywhere, so I have removed the figure but simply put 0.1% as first-party results for mobile. If you'd like to have a look - here's the data - https://docs.google.com/spreadsheets/d/1zU9rHpI3nC6jTz3xgN6w13afW7x34xAKBh2IPH-lVxk/edit#gid=2057978707 #}

Author: @tunetheweb and @NishuGoel  this seems to be leftovers from before

Reviewer 3: We discussed after. Happy to remove this TODO. I was wrong 😁

        <td class="numeric">1.3%</td>
      </tr>
      <tr>
        <td>AngularJS</td>
        <td class="numeric">1.0%</td>
      </tr>
      <tr>
        <td>Mustache</td>
        <td class="numeric">0.7%</td>
      </tr>
      <tr>
        <td>jQuery Mobile</td>
        <td class="numeric">0.5%</td>
      </tr>
      <tr>
        <td>Dojo</td>
        <td class="numeric">0.5%</td>
      </tr>
      <tr>
        <td>Angular</td>
        <td class="numeric">0.4%</td>
      </tr>
        <td>Vue</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>Knockout</td>
        <td class="numeric">0.2%</td>
      </tr>
      <tr>
        <td>Highcharts</td>
        <td class="numeric">0.1%</td>
      </tr>
      <tr>
        <td>Next.js</td>
        <td class="numeric">0.0%</td>
      </tr>
      <tr>
        <td>React</td>
        <td class="numeric">0.0%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(
    caption="The percent of mobile pages found to contain a vulnerable version of a JavaScript library or framework.",
    sheets_gid="551476210",
    sql_file="lighthouse_vulnerable_libraries.sql"
  ) }}</figcaption>
</figure>

> When we segment the percent of mobile pages by library and framework, we can see that jQuery is largely responsible for the decrease vulnerabilities. This year JavaScript vulnerabilities were found on 57.6% of pages with jQuery, compared to [80.9% last year](../2020/javascript#fig-31). As [predicted](../2020/javascript#fig-31) by [Tim Kadlec](../2020/contributors#tkadlec) in the 2020 edition of this chapter, <em>"if we can get folks to migrate away from those outdated, vulnerable versions of jQuery, we would see the number of sites with known vulnerabilities plummet"</em>. And that's exactly what happened; WordPress migrated from jQuery version 1.12.4 to the more secure version 3.5.1, contributing to a 20 point drop in the percent of pages with known JavaScript vulnerabilities.

Reviewer 2: ```suggestion
When we segment the percent of mobile pages by library and framework, we can see that jQuery is largely responsible for the decrease in vulnerabilities. This year JavaScript vulnerabilities were found on 57.6% of pages with jQuery, compared to [80.9% last year](../2020/javascript#fig-31). As [predicted](../2020/javascript#fig-31) by [Tim Kadlec](../2020/contributors#tkadlec) in the 2020 edition of this chapter, <em>"if we can get folks to migrate away from those outdated, vulnerable versions of jQuery, we would see the number of sites with known vulnerabilities plummet"</em>. And that's exactly what happened; WordPress migrated from jQuery version 1.12.4 to the more secure version 3.5.1, contributing to a 20 point drop in the percent of pages with known JavaScript vulnerabilities.
```

## How do we use JavaScript?

Now that we've looked at how we get the JavaScript, what are we using it for?

### AJAX

A lot of JavaScript that is written is used to communicate with servers through asynchronous requests to receive information in various formats. AJAX, or asynchronous JavaScript and XML, is typically used to send and receive data in JSON, XML, HTML, text formats.

> {# TODO - "AJAX, or asynchronous JavaScript and XML," is this a list of 3 or grouped together? #}

Author: @NishuGoel

Reviewer: full form of AJAX so group together, not a list of 3

Reviewer 3: Would such we flip it around and put in italics to show it's introducing a new term:

```
_Asynchronous JavaScript and XML_ (AJAX), is typically used to send and receive data in JSON, XML, HTML, text formats.

  sheets_gid="1012977216",
  sql_file="percentage_usage_of_different_ajax_apis.sql"
  )
}}

In the case of the usage of the `Fetch` API, the median mobile page makes 2 requests, and in the long tail, reaches 3 requests. This API is becoming the standard XHR way of making requests, due in part to its cleaner approach and less boilerplate code. There may also be <a hreflang="en" href="https://gomakethings.com/the-fetch-api-performance-vs.-xhr-in-vanilla-js/">performance benefits</a> to `Fetch` over the traditional XHR approach, due to the way browsers can decode large JSON payloads off the main thread.

{{ figure_markup(
  image="ajax_beacon.png",
  caption="Distribution of the number of `Beacon` requests per page.",
  description="Bar chart showing the usage of Beacon per page on both desktop and mobile pages. The median mobile page makes no Beacon requests, but at 90th percentile, makes 1 Beacon request.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTpHzC_cMZYj2VLzQ4ODK3uvZkNBXtwdAZriZaBwjLjUM1SGwwmJs9rv8T6OtNdXox29PQ34CasUUwc/pubchart?oid=1483701430&format=interactive",
  sheets_gid="1012977216",
  sql_file="percentage_usage_of_different_ajax_apis.sql"
  )
}}

> `Beacon` usage is almost non-existant, with 0 requests per page until the 90th percentile, at which there's only one request per page. One possible explanation for this low adoption could be that `Beacon` is typically used for sending analytics data, especially when one wants to ensure that the request is sent even if the page might unload soon. This is, however, not guaranteed when using XHR. A good experiment for the future would be to see if some statistics could be collected around any pages using XHR for analytics data, session data, etc.

Reviewer 2: ```suggestion
`Beacon` usage is almost non-existent, with 0 requests per page until the 90th percentile, at which there's only one request per page. One possible explanation for this low adoption could be that `Beacon` is typically used for sending analytics data, especially when one wants to ensure that the request is sent even if the page might unload soon. This is, however, not guaranteed when using XHR. A good experiment for the future would be to see if some statistics could be collected around any pages using XHR for analytics data, session data, etc.
```

It is observed that so many features provided for improving the rendering time and resource loading time could be leveraged better to see the overall impact and get an even better experience with respect to the performance.

Starting from adopting the features provided by the JavaScript and web community, making sure to use them wisely and ensuring that a tweak only improves the performance and doesn't instead increase the overload (for example, lazy loading all resources on a page is definitely not a good idea), is what it will take us to see these numbers even better in the coming years.

> {# TODO - this sentence is long but not sure how to break it down #}

Reviewer 3: Agreed and not sure that this is trying to say. Does this cover it?

> Start by adopting the new features provided by the JavaScript and web community. However, make sure to use them wisely and ensure that they improve performance and don't instead increase the load, as some APIs can cause harm through overused. For example, lazy loading should be applied selectively and not bluntly set on all resources on a page. Making appropriate use of the powerful APIs we now have access to, is what it will take to see these numbers improve further in the coming years.

FYI @shantsis if you select all three lines when making comments than people can suggest fixes in these comments and commit straight away from these comments. At the moment I can only suggest changes to the line you've commented on (the TODO bit).

> {# TODO - this sentence is long but not sure how to break it down #}

Reviewer: Yup this covers what I was trying to say. 
"through overused"? - do you mean "through overuse"?

> {# TODO - this sentence is long but not sure how to break it down #}

Reviewer 3: Yup. Edited it above.

Is lazy loading the best example though? Especially with native lazy loading now so common, using JS for this is less necessary now (I argue not necessary at all!). Wonder if there's a better example we could use here that's mroe relatable to JS?

> {# TODO - this sentence is long but not sure how to break it down #}

Author: thanks for the tip @tunetheweb!

> {# TODO - this sentence is long but not sure how to break it down #}

Reviewer: >Wonder if there's a better example we could use here that's more relatable to JS?

How about async/defer?
defer 's usage becoming meaningless in the presence of async.

> {# TODO - this sentence is long but not sure how to break it down #}

Reviewer 3: Better example!

> Start by adopting the new features provided by JavaScript and the web community. However, make sure to use them wisely and ensure that they improve performance and don't instead increase the load, as some APIs can cause harm through overuse. For example, we saw that using `async` as well as `defer` is very common and actually is an anti-pattern that slows the site down since the `defer` is effectively ignored. Making appropriate use of the powerful APIs we now have access to, is what it will take to see these numbers improve further in the coming years.
